doc/adr/0016-end-to-end-testing-strategy.rst

```diff
@@ -0,0 +1,277 @@
+.. _end_to_end_testing_strategy:
+
+ADR 16: End-to-end testing strategy for Plutus, cardano-ledger-api and cardano-node-client
+==========================================================================================
+
+Date: 2022-12-02
+
+Author(s)
+---------
+
+james-iohk <james.browning@iohk.io>
+
+Status
+------
+
+Draft
+
+Context
+-------
+
+End-to-end testing of `Plutus Core <https://github.com/input-output-hk/Plutus/>`_ functionality is
+currently performed by a combination of automation and exploratory approaches, both are performed
+on public `preview` and `pre-prod` testnets using a real node.
+Automation test scenarios for Plutus are currently being run as part of the wider `cardano-node-tests
+<https://github.com/input-output-hk/cardano-node-tests/>`_ test suite, which uses a Python wrapper
+for ``cardano-cli``.
+Those tests focus on general ledger/node/cli functionality and only cover a few key scenarios for
+Plutus functionality, such as TxInfo and SECP256k1
+builtins.
+
+There is also ongoing development work to separate the functionality of `cardano-api
+<https://github.com/input-output-hk/cardano-node/tree/master/cardano-api>`_ out into two packages:
+
+* `cardano-ledger-api
+  <https://github.com/input-output-hk/cardano-ledger/tree/master/libs/cardano-ledger-api>`_ handles
+  the building and balancing of transactions.
+
+* ``cardano-node-client`` will live in `cardano-node
+  <https://github.com/input-output-hk/cardano-node>`_ and handle the submitting of balanced
+  transactions and querying the ledger state.
+
+Both of these packages are in early stage development and will require end-to-end test coverage.
+
+This document outlines the decisions and arguments for an **additional** approach to end-to-end test
+automation using a framework written in Haskell.
+
+The exploratory testing approach is not in the scope of this document.
```

Comment:
In preparation for major node release (hard-fork), in the phase where plutus/ledger/node/cli are being integrated,  I'll spend some days/weeks hands-on with the node-cli experimenting with new features and how they behave in combination with the existing. It's a good way to feel around and provide quick feedback (bugs/ux) to devs and product, especially when expectations are incomplete.

The problem with it currently is automated tests don't exist until days/weeks after so there lots of manual checks for regression during the final stages of development. It would be good to have at least some key end-to-end scenarios ready to assist, or even cover most the work.

It's not something I believe we can replace entirely but can be made more efficient.

Reply:
Exploratory testing is top-down, being a user with an outside perspective, trying things you're not supposed to and sometimes stumbling upon the unknown. Most useful when testing interfaces designed for humans, impossible to automate. I find it inspiring.